### PR TITLE
Add last updated date / reviewed date to FxOS Add-ons (bug 1205791)

### PR DESCRIPTION
**Pull request by diox**
https://bugzilla.mozilla.org/show_bug.cgi?id=1205791

### CODE REVIEW

**Comment by diox** [comment]
In the bug :
>> We can't just use addon.latest_vesion.date_modified?
>
> Mmmm, we could. But then as soon as we introduce a new field that the developer can edit on `ExtensionVersion` we'll need to denormalize. I kinda prefer doing it now, it's not very costly and makes sense.

File: mkt/extensions/indexers.py [hunk 1]
```diff
@@ -113,7 +120,9 @@ def extract_document(cls, pk=None, obj=None):
         else:
             doc['latest_public_version'] = None
         doc['name_sort'] = unicode(obj.name).lower()
-        doc['guid'] = unicode(obj.uuid)
+        # Find the first reviewed date (used in sort).
```
Comment (chuckharmston) [inline mkt/extensions/indexers.py]: Is the sort here "first time approved in Marketplace"? That's not particularly intuitive.

File: mkt/extensions/models.py [hunk 2]
```diff
@@ -52,10 +53,11 @@ def public(self):
 
 class Extension(ModelBase):
     # Automatically handled fields.
-    uuid = UUIDField(auto=True)
+    last_updated = models.DateTimeField(db_index=True, null=True)
```
Comment (chuckharmston) [inline mkt/extensions/models.py]: Why is this nullable? Seems to me we'd want `default=timezone.now` instead.

File: mkt/extensions/models.py [hunk 3]
```diff
@@ -359,9 +363,13 @@ def is_version_number_higher(cls, version1, version2):
         return v1 > v2
 
     def publish(self):
-        """Publish this extension version to public."""
+        """Publish this extension version to public.
```
Comment (chuckharmston) [inline mkt/extensions/models.py]: This sentence is pretty awkward. Some suggestions:

1. Make this extension version public.
2. Change this extension version's status to public.
3. Publicize this extension version.

File: mkt/extensions/tests/test_indexers.py [hunk 4]
```diff
@@ -12,11 +14,17 @@ class TestExtensionIndexer(TestCase):
     def setUp(self):
         self.indexer = Extension.get_indexer()()
 
-    def _extension_factory(self, status=STATUS_PENDING):
+    def _extension_factory(self, status=STATUS_PENDING, reviewed=None):
+        if reviewed is None:
+            # Microseconds are not saved by MySQL, so set it to 0 to make sure
+            # our comparisons still work once the model is saved to the db.
```
Comment (chuckharmston) [inline mkt/extensions/tests/test_indexers.py]: Good catch. I'm surprised we don't have a helper for this somewhere.

File: mkt/extensions/tests/test_views.py [hunk 5]
```diff
@@ -148,6 +149,11 @@ def setUp(self):
         self.list_url = reverse('api-v2:extension-list')
         self.user = UserProfile.objects.get(pk=2519)
 
+    def tearDown(self):
+        super(TestExtensionViewSetPost, self).tearDown()
+        # Explicitely delete the Extensions to clean up leftover files.
```
Comment (chuckharmston) [inline mkt/extensions/tests/test_views.py]: With a quick skim of this commit, I don't see anywhere where you'd be creating files. Should this be there already?

Comment (chuckharmston) [inline mkt/extensions/tests/test_views.py]: In general we should be mocking things instead of using the filesystem, anyway.

File: docs/api/topics/firefox_os_addons.rst [hunk 6]
```diff
@@ -67,6 +68,7 @@ Detail
 
     :resjson int id: The add-on id.
     :resjson string|object|null description: The add-on description.
+    :resjson string|null last_updated: The latest date a version was published at for this add-on.
```
Comment (chuckharmston) [inline docs/api/topics/firefox_os_addons.rst]: What about `reviewed`?

**Comment by chuckharmston** [comment]
r+wc

File: mkt/extensions/tests/test_views.py (hunk 5 above)
Comment (diox) [inline mkt/extensions/tests/test_views.py]: Yeah, sorry, this is a drive-by fix. I stumbled upon it when re-launching the tests locally a bunch of times when writing this PR, but it was caused by earlier commits.

File: mkt/reviewers/tests/test_views.py [hunk 7]
```diff
@@ -1249,7 +1249,7 @@ def test_public_sign(self, sign_mock, json_mock, update_cached_manifests):
         with private_storage.open(
                 self.version.files.all()[0].file_path, 'w') as f:
             f.write('.')
-        public_storage.delete(self.version.files.all()[0].file_path)
+        public_storage.delete(self.version.files.all()[0].signed_file_path)
```
Comment (diox) [inline mkt/reviewers/tests/test_views.py]: Also a drive-by fix.

File: mkt/files/models.py [hunk 8]
```diff
@@ -227,7 +227,7 @@ def cleanup_file(sender, instance, **kw):
     if kw.get('raw') or not instance.filename:
         return
     # Use getattr so the paths are accessed inside the try block.
-    for path in ('file_path', 'guarded_file_path'):
+    for path in ('file_path', 'guarded_file_path', 'signed_file_path'):
```
Comment (diox) [inline mkt/files/models.py]: drive-by fix.

File: mkt/extensions/tests/test_views.py (hunk 5 above)
Comment (diox) [inline mkt/extensions/tests/test_views.py]: Yeah, but for the upload tests it's kinda painful to fully mock the filesystem so I didn't do it. I did it for the rest of the tests though, like reject/publish.

File: mkt/extensions/indexers.py (hunk 1 above)
Comment (diox) [inline mkt/extensions/indexers.py]: It's the sort used by "new". Updates are not displayed there, only new things, ordered by review date.

File: mkt/extensions/models.py (hunk 2 above)
Comment (diox) [inline mkt/extensions/models.py]: We don't, because it's only set when a new version is `*published*`. We don't want it to be updated on other saves.

File: mkt/extensions/indexers.py (hunk 1 above)
Comment (chuckharmston) [inline mkt/extensions/indexers.py]: Huh. I'd expect things to be bumped by updates.

¯\_(ツ)_/¯

File: docs/api/topics/firefox_os_addons.rst (hunk 6 above)
Comment (diox) [inline docs/api/topics/firefox_os_addons.rst]: It does not need to be exposed right now.

File: mkt/extensions/indexers.py (hunk 1 above)
Comment (diox) [inline mkt/extensions/indexers.py]: Yeah. I suspect we did that in order to avoid having developers constantly push new updates to stay at the top of the new page (they'd need reviews, but still)
